elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java

```diff
@@ -95,14 +117,78 @@ else if (queryObj.getQueryType().equals(QueryType.GRAPHQL_V1_0)) {
 
         queryResultObj = new AsyncQueryResult();
         queryResultObj.setHttpStatus(response.getResponseCode());
-        queryResultObj.setResponseBody(response.getBody());
         queryResultObj.setContentLength(response.getBody().length());
-        queryResultObj.setResultType(ResultType.EMBEDDED);
+        queryResultObj.setRecordCount(recCount);
         queryResultObj.setCompletedOn(new Date());
 
+        URL url = null;
+        String tempResult = response.getBody();
+        if (queryObj.getResultFormatType() == ResultFormatType.CSV) {
+            tempResult = convertJsonToCSV(response.getBody());
+        }
+        byte[] temp = tempResult.getBytes();
+
+        if (queryObj.getResultType() == ResultType.DOWNLOAD) {
+            url = resultStorageEngine.storeResults(queryObj.getId(), temp);
+            queryResultObj.setResponseBody(url.toString());
+        } else if (queryObj.getResultType() == ResultType.EMBEDDED) {
+            queryResultObj.setResponseBody(tempResult);
+        }
+
         return queryResultObj;
     }
 
+    /**
+     * This method calculates the number of records from the response with JSON API.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return rec is the recordCount
+     */
+    protected Integer calculateRecordsJSON(String jsonStr) {
+        Integer rec = null;
+        rec = JsonPath.read(jsonStr, "$.data.length()");
+        return rec;
+    }
+
+    /**
+     * This method calculates the number of records from the response with GRAPHQL API.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return rec is the recordCount
+     */
+    protected Integer calculateRecordsGraphQL(String jsonStr) throws IOException {
+        Integer rec = null;
+
+        Configuration conf = Configuration.builder()
+                .options(Option.AS_PATH_LIST).build();
+
+        List<String> pathList = using(conf).parse(jsonStr).read("$..edges");
+        String modify = pathList.get(0).replace("[", ".");
```

Comment:
L-164 to 166 can be changed as
`pathList.get(0).replace("[", ".").replace("'", "").replace("]", "") + ".length()"`